src/ir_pattern.py

```diff
@@ -0,0 +1,465 @@
+from .parse_instruction import (
+    ArbitraryMemoryLocation,
+    Argument,
+    AsmAddressMode,
+    AsmGlobalSymbol,
+    AsmInstruction,
+    AsmLiteral,
+    BinOp,
+    Instruction,
+    InstructionMeta,
+    JumpTarget,
+    LocalLocation,
+    Location,
+    Macro,
+    MemoryLocation,
+    RegFormatter,
+    Register,
+    StackLocation,
+    make_location,
+    parse_instruction,
+)
+
+
+@dataclass(eq=False, frozen=True)
+class IrPattern(abc.ABC):
+    """
+    Template for defining "IR" patterns that can match against input asm.
+    The matching process uses the FlowGraph and register analysis to compute
+    inter-instruction dependencies, so these patterns can match even when
+    they have been interleaved/reordered by the compiler in the input asm.
+
+    IrPattern subclasses *must* define `parts` and `replacement`, and can
+    optionally implement `check()`.
+
+    For now, the pattern cannot contain any branches, and the replacement
+    must be a single instruction (though, it can be fictive).
+    """
+
+    parts: ClassVar[List[str]]
+    replacement: ClassVar[str]
+
+    flow_graph: FlowGraph
+    replacement_instr: Instruction
+
+    def check(self, m: "IrMatch") -> bool:
+        """Override to perform additional checks/calculations before replacement."""
+        return True
+
+    @classmethod
+    def compile(cls, arch: ArchFlowGraph) -> "IrPattern":
+        missing_meta = InstructionMeta.missing()
+        regf = RegFormatter()
+        replacement_instr = parse_instruction(cls.replacement, missing_meta, arch, regf)
+
+        name = f"__pattern_{cls.__name__}"
+        func = Function(name=name, arguments=[])
+        # Add a fictive nop instruction for each input to the replacement_instr
+        # This acts as a placeholder Reference to represent where the input was set
+        for inp in replacement_instr.inputs:
+            func.new_instruction(
+                Instruction(
+                    "in.fictive",
+                    [],
+                    meta=missing_meta,
+                    inputs=[],
+                    clobbers=[],
+                    outputs=[inp],
+                )
+            )
+        for part in cls.parts:
+            func.new_instruction(parse_instruction(part, missing_meta, arch, regf))
+        # Add a fictive nop instruction for each output from the replacement_instr
+        for out in replacement_instr.outputs:
+            func.new_instruction(
+                Instruction(
+                    "out.fictive",
+                    [],
+                    meta=missing_meta,
+                    inputs=[out],
+                    clobbers=[],
+                    outputs=[],
+                )
+            )
+
+        asm_data = AsmData()
+        flow_graph = build_flowgraph(func, asm_data, arch, fragment=True)
+        return cls(
+            flow_graph=flow_graph,
+            replacement_instr=replacement_instr,
+        )
+
+
+@dataclass
+class IrMatch:
```

Comment:
I could experiment with trying to share some code between the two once the rest of the PR looks good? I think it'd get too confusing if I try to refactor it now.

Reply:
For posterity: I experimented with reducing duplication on this branch: https://github.com/zbanks/mips_to_c/compare/zbanks:ir-patterns...zbanks:unify-ir-asm-match

though I wasn't very happy with this refactor because it involved a lot of weird OOP tricks.